### PR TITLE
Minimize dependencies on trait and infer inside librustc

### PR DESCRIPTION
**Pull request by cjgillot**
Split from #67953 

All commits should pass check on their own.

r? @Centril

### CODE REVIEW

**Comment by cjgillot** [comment]
Rebased.

File: src/librustc/traits/drop.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
+//! Miscellaneous type-system utilities that are too small to deserve their own modules.
```
Comment (Centril) [inline src/librustc/traits/drop.rs]: The module name is a bit misleading (since it's not just about `Drop`); I'd name it `misc.rs` or something.

**Review by Centril — COMMENTED** [review]
Thanks!

The comment below is the only thing I'd prefer having addressed now before this can land.

However, I also had some other thoughts:

- Moving `ty::wf` to traits feels odd because well-formedness of types isn't a semantic property of traits only.

- The `free_region_map` should perhaps not be in `ty`.

That said, I think we can experiment for now and reassess once more things like `Ty<'tcx>` has been moved out. Perhaps more opportunities for good semantic categorization will arise then.

Also, let's cc @nikomatsakis & @matthewjasper for awareness. :)

**Comment by matthewjasper** [comment]
>     * The `free_region_map` should perhaps not be in `ty`.

It shouldn't. It's currently in `TypeckTables` but it's unused there.

**Pull request merged**